File: pointer/#680_valid_palindromeII/validPalindrome.go
package main

func main() {
	validPalindrome("deeee")                  // true
	validPalindrome("eeccccbebaeeabebccceea") // false

	// validPalindrome("aba")
	// validPalindrome("abca")

}

// Recursive
func validPalindrome(s string) bool {
	l := 0
	r := len(s) - 1

	if s[l] == s[r] {
		return isValidPalindrome(s, l+1, r-1, false)
	}

	return isValidPalindrome(s, l+1, r, true) || isValidPalindrome(s, l, r-1, true)
}

func isValidPalindrome(s string, l, r int, isRemoved bool) bool {
	if l >= r {
		return true
	}

	if s[l] == s[r] {
		return isValidPalindrome(s, l+1, r-1, isRemoved)
	}

	if isRemoved {
		return false
	}

	return isValidPalindrome(s, l+1, r, true) || isValidPalindrome(s, l, r-1, true)
}

// Iterative
func validPalindrome2(s string) bool {
	l := 0
	r := len(s) - 1
	for l < r {
		if s[l] != s[r] {
			return isPalindrome(s, l+1, r) || isPalindrome(s, l, r-1)
		}
		l++
		r--
	}

	return true
}

func isPalindrome(s string, l, r int) bool {
	for l < r {
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}

	return true
}

File: pointer/#680_valid_palindromeII/validPalindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var validPalindrome = function (s) {
  let l = 0,
    r = s.length - 1;
  while (l < r) {
    // if at any point the characters don't match,
    // then check once from left to right - 1,
    // and from left + 1 to right,
    // if for both cases find false, return false else return true.
    if (s[l] !== s[r]) {
      return isPalindrome(s, l, r - 1) | isPalindrome(s, l + 1, r);
    }

    l++;
    r--;
  }

  return true;
};

var isPalindrome = function (s, start, end) {
  while (start < end) {
    if (s[start] !== s[end]) {
      return false;
    }

    start++;
    end--;
  }

  return true;
};

validPalindrome("aba"); // true
validPalindrome("abca"); // true
validPalindrome("abc"); // false
validPalindrome("eeccccbebaeeabebccceea"); // false
validPalindrome("deeee"); // true
